test/test_ekfac.py

```diff
@@ -0,0 +1,949 @@
+"""Contains tests for ``EKFACLinearOperator`` in ``curvlinops.kfac``."""
+
+from typing import Dict, Iterable, List, Tuple, Union
+
+from einops.layers.torch import Rearrange
+from pytest import mark, raises
+from torch import (
+    Tensor,
+    allclose,
+    device,
+    eye,
+    linalg,
+    manual_seed,
+    rand,
+    rand_like,
+)
+from torch.nn import (
+    BCEWithLogitsLoss,
+    CrossEntropyLoss,
+    Linear,
+    Module,
+    MSELoss,
+    Parameter,
+    Sequential,
+)
+
+from curvlinops import EFLinearOperator, FisherMCLinearOperator, GGNLinearOperator
+from curvlinops.ekfac import EKFACLinearOperator, FisherType, KFACType
+from curvlinops.kfac import KFACLinearOperator
+from curvlinops.utils import allclose_report
+from test.cases import DEVICES, DEVICES_IDS
+from test.utils import (
+    Conv2dModel,
+    UnetModel,
+    WeightShareModel,
+    _test_inplace_activations,
+    _test_property,
+    _test_save_and_load_state_dict,
+    binary_classification_targets,
+    block_diagonal,
+    classification_targets,
+    compare_state_dicts,
+    maybe_exclude_or_shuffle_parameters,
+    regression_targets,
+)
+
+
+@mark.parametrize(
+    "separate_weight_and_bias", [True, False], ids=["separate_bias", "joint_bias"]
+)
+@mark.parametrize(
+    "exclude", [None, "weight", "bias"], ids=["all", "no_weights", "no_biases"]
+)
+@mark.parametrize("shuffle", [False, True], ids=["", "shuffled"])
+def test_ekfac_type2(
+    kfac_exact_case: Tuple[
+        Module, MSELoss, List[Parameter], Iterable[Tuple[Tensor, Tensor]]
+    ],
+    shuffle: bool,
+    exclude: str,
+    separate_weight_and_bias: bool,
+):
+    """Test the EKFAC implementation against the exact GGN.
+
+    Args:
+        kfac_exact_case: A fixture that returns a model, loss function, list of
+            parameters, and data.
+        shuffle: Whether to shuffle the parameters before computing the EKFAC matrix.
+        exclude: Which parameters to exclude. Can be ``'weight'``, ``'bias'``,
+            or ``None``.
+        separate_weight_and_bias: Whether to treat weight and bias as separate blocks in
+            the EKFAC matrix.
+    """
+    model, loss_func, params, data, batch_size_fn = kfac_exact_case
+    params = maybe_exclude_or_shuffle_parameters(params, model, exclude, shuffle)
+
+    ggn = block_diagonal(
+        GGNLinearOperator,
+        model,
+        loss_func,
+        params,
+        data,
+        batch_size_fn=batch_size_fn,
+        separate_weight_and_bias=separate_weight_and_bias,
+        return_numpy=False,
+    )
+    ekfac = EKFACLinearOperator(
+        model,
+        loss_func,
+        params,
+        data,
+        batch_size_fn=batch_size_fn,
+        fisher_type=FisherType.TYPE2,
+        separate_weight_and_bias=separate_weight_and_bias,
+    )
+    ekfac_mat = ekfac @ eye(ekfac.shape[1])
```

Comment:
This should crash if `ekfac` lives on GPU.

Reply:
There are a couple more of these cases in this file.

Reply:
I extended `eye_like` to also work with linops and use it everywhere in this file. In the future when we remove `to_scipy()` calls elsewhere we can also make use of it.